Failed compilation with GCC 9 or newer

buildroot's gcc 9+ failed to compile OCC.

### Issue 1:

```
..x/ppc32 -I/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/ssx/ppc405 -I/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/ssx/trace -I/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/ssx/occhw -I/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/ssx/../lib/common -I/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/ssx/../include -I/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/ssx/../include/registers -I/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/lib/occlib -I/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/lib/common -I/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/lib/ssxlib -I/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/lib/ppc405lib -DTRAC_TO_SIMICS=0 -DSIMICS_ENVIRONMENT=0 -DIMAGE_NAME=occ_405 -DSSX_TIMER_SUPPORT=1 -DSSX_THREAD_SUPPORT=1 -DPPC405_MMU_SUPPORT=0 -DSSX_TRACE_SUPPORT=1 -DSSX_TRACE_HASH_PREFIX=38241 -DSSX_USE_INIT_SECTION=0 -DUSE_SSX_APP_CFG_H=1 -D__SSX__=1 -DSSX_NO_BOOT_VECTOR0=1 -o /home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/obj/occ_405/cent/centaur_control.o cent/centaur_control.c
powerpc64le-buildroot-linux-gnu-gcc.br_real: warning: switch ‘-mstring’ is no longer supported
powerpc64le-buildroot-linux-gnu-gcc.br_real: warning: switch ‘-mstring’ is no longer supported
cent/centaur_control.c: In function ‘centaur_control_init’:
cent/centaur_control.c:389:17: error: taking address of packed member of ‘struct memory_control_task’ may result in an unaligned pointer value [-Werror=address-of-packed-member]
  389 |                 &G_memory_control_task.gpe_req,            // gpe_req for the task
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
```

### Issue 2:

```
E_INIT_SECTION=0 -DUSE_SSX_APP_CFG_H=1 -D__SSX__=1 -DSSX_NO_BOOT_VECTOR0=1 -o /home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/obj/occ_405/cent/centaur_data.o cent/centaur_data.c
powerpc64le-buildroot-linux-gnu-gcc.br_real: warning: switch ‘-mstring’ is no longer supported
powerpc64le-buildroot-linux-gnu-gcc.br_real: warning: switch ‘-mstring’ is no longer supported
cent/centaur_data.c: In function ‘centaur_data’:
cent/centaur_data.c:645:37: error: taking address of packed member of ‘struct membuf_data_task’ may result in an unaligned pointer value [-Werror=address-of-packed-member]
  645 |         if( !(async_request_is_idle(&l_centaur_data_ptr->gpe_req.request)) )
      |                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cent/centaur_data.c:684:41: error: taking address of packed member of ‘struct membuf_data_task’ may result in an unaligned pointer value [-Werror=address-of-packed-member]
  684 |             if(!async_request_completed(&l_centaur_data_ptr->gpe_req.request) || l_parms->error.rc )
      |                                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cent/centaur_data.c:889:40: error: taking address of packed member of ‘struct membuf_data_task’ may result in an unaligned pointer value [-Werror=address-of-packed-member]
  889 |             rc = gpe_request_schedule( &(l_centaur_data_ptr->gpe_req) );
      |                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cent/centaur_data.c: In function ‘centaur_init’:
cent/centaur_data.c:1196:17: error: taking address of packed member of ‘struct membuf_data_task’ may result in an unaligned pointer value [-Werror=address-of-packed-member]
 1196 |                 &G_membuf_data_task.gpe_req,     //gpe_req for the task
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
error from cc
cc returned 1
```

### Issue 3:

```
make[3]: Leaving directory '/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/lib/common'
/home/tle/Work/op-build/output/host/bin/powerpc64le-buildroot-linux-gnu-ld -e __ssx_boot -T/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/obj/occ_405/linkscript --oformat=elf32-powerpc -melf32ppc -Map /home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/obj/occ_405/occ_405.map -Bstatic -o /home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/obj/occ_405/occ_405.out -L/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/obj/occ_405 -L/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/obj/occ_405/ssx -L/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/obj/occ_405/ssx/ppc405 -L/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/obj/occ_405/ssx/ppc32 -L/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/obj/occ_405/commonlib -L/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/obj/occ_405/occlib -L/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/obj/occ_405/ppc405lib -L/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/obj/occ_405/cmdh -L/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/obj/occ_405/dcom -L/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/obj/occ_405/dimm -L/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/obj/occ_405/errl -L/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/obj/occ_405/gpu -L/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/obj/occ_405/lock -L/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/obj/occ_405/pss -L/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/obj/occ_405/rtls -L/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/obj/occ_405/sensor -L/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/obj/occ_405/thread -L/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/obj/occ_405/timer -L/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/obj/occ_405/trac -L/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/obj/occ_405/amec -L/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/obj/occ_405/dcom -L/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/obj/occ_405/proc -L/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/obj/occ_405/firdata -L/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/obj/occ_405/cent -L/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/obj/occ_405/mem -L/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/obj/occ_405/wof -L/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/obj/occ_405/pgpe -lssx -locc -lppc405 -lcommon
/home/tle/Work/op-build/output/host/bin/powerpc64le-buildroot-linux-gnu-ld: /home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/obj/occ_405/main.o: in function `mainThrdTimerCallback':
/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/occ_405/main.c:1495: undefined reference to `__stack_chk_fail'
/home/tle/Work/op-build/output/host/bin/powerpc64le-buildroot-linux-gnu-ld: /home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/obj/occ_405/main.o: in function `read_wof_header':
/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/occ_405/main.c:509: undefined reference to `__stack_chk_fail'
/home/tle/Work/op-build/output/host/bin/powerpc64le-buildroot-linux-gnu-ld: /home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/obj/occ_405/main.o: in function `read_pgpe_header':
/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/occ_405/main.c:714: undefined reference to `__stack_chk_fail'
/home/tle/Work/op-build/output/host/bin/powerpc64le-buildroot-linux-gnu-ld: /home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/obj/occ_405/main.o: in function `read_ppmr_header':
/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/occ_405/main.c:841: undefined reference to `__stack_chk_fail'
/home/tle/Work/op-build/output/host/bin/powerpc64le-buildroot-linux-gnu-ld: /home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/obj/occ_405/main.o: in function `read_oppb_params':
/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/occ_405/main.c:1020: undefined reference to `__stack_chk_fail'
/home/tle/Work/op-build/output/host/bin/powerpc64le-buildroot-linux-gnu-ld: /home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/obj/occ_405/main.o:/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/occ_405/main.c:1254: more undefined references to `__stack_chk_fail' follow
make[2]: *** [Makefile:87: /home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/obj/occ_405/occ_405.out] Error 1
```

### Issue 4

```
86e21a/src/occ_405/pss -I/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/occ_405/timer -I/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/occ_405/proc -I/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/occ_405/cmdh -I/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/occ_405/dcom -I/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/occ_405/amec -I/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/occ_405/cent -I/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/occ_405/dimm -I/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/occ_405/gpu -I/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/occ_405/mem -I/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/occ_405/lock -I/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/occ_405/wof -I/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/occ_405/../common -I/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/occ_405/pgpe  -I/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/ssx/ssx -I/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/ssx/ppc32 -I/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/ssx/ppc405 -I/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/ssx/trace -I/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/ssx/occhw -I/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/ssx/../lib/common -I/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/ssx/../include -I/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/ssx/../include/registers -I/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/lib/occlib -I/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/lib/common -I/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/lib/ssxlib -I/home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/src/lib/ppc405lib -DTRAC_TO_SIMICS=0 -DSIMICS_ENVIRONMENT=0 -DIMAGE_NAME=occ_405 -DSSX_TIMER_SUPPORT=1 -DSSX_THREAD_SUPPORT=1 -DPPC405_MMU_SUPPORT=0 -DSSX_TRACE_SUPPORT=1 -DSSX_TRACE_HASH_PREFIX=38241 -DSSX_USE_INIT_SECTION=0 -DUSE_SSX_APP_CFG_H=1 -D__SSX__=1 -DSSX_NO_BOOT_VECTOR0=1 -o /home/tle/Work/op-build/output/build/occ-16131c38c2e593206447403c3ac79b18fb86e21a/obj/occ_405/wof/wof.o wof/wof.c
powerpc64le-buildroot-linux-gnu-gcc.br_real: warning: switch ‘-mstring’ is no longer supported
powerpc64le-buildroot-linux-gnu-gcc.br_real: warning: switch ‘-mstring’ is no longer supported
wof/wof.c: In function ‘calculate_core_voltage’:
wof/wof.c:1061:51: error: array subscript ‘OCCPstateTable_entry_t[0]’ is partly outside array bounds of ‘uint32_t[1]’ {aka ‘unsigned int[1]’} [-Werror=array-bounds]
 1061 |             uint8_t current_vid = pstate_entry_ptr->internal_vdd_vid;
      |                                                   ^~
wof/wof.c:1057:22: note: while referencing ‘current_pstate’
 1057 |             uint32_t current_pstate = in32(pstate_addr);
```

--- a/src/occBootLoader/img_defs.mk
+++ b/src/occBootLoader/img_defs.mk
@@ -227,7 +227,7 @@ PIPE-CFLAGS = -pipe -Wa,-m405
 GCC-CFLAGS += -g -Wall -fsigned-char -msoft-float  \
 	-m32 -mbig-endian -mcpu=405 -mmultiple -mstring \
 	-meabi -msdata=eabi -ffreestanding -fno-common \
-	-fno-inline-functions-called-once \
+	-fno-inline-functions-called-once -fno-pic \
 	-fno-asynchronous-unwind-tables -std=gnu89
 
 CFLAGS      =  -c $(GCC-CFLAGS) $(PIPE-CFLAGS) $(GCC-O-LEVEL) $(INCLUDES)

--- a/src/occ_405/img_defs.mk
+++ b/src/occ_405/img_defs.mk
@@ -255,6 +255,9 @@ GCC-CFLAGS += -g -Wall -fsigned-char -msoft-float  \
 	-m32 -mbig-endian -mcpu=405 -mmultiple -mstring \
 	-meabi -msdata=eabi -ffreestanding -fno-common \
 	-fno-inline-functions-called-once \
+	-fno-pic -fno-stack-protector \
+	-Wno-address-of-packed-member \
+	-Wno-array-bounds \
 	-fno-asynchronous-unwind-tables -std=gnu89
 
 CFLAGS      =  -c $(GCC-CFLAGS) $(PIPE-CFLAGS) $(GCC-O-LEVEL) $(INCLUDES)

